pkg/query/querier.go

```diff
@@ -345,8 +358,9 @@ func (q *querier) selectFn(ctx context.Context, hints *storage.SelectHints, ms .
 		}, nil
 	}
 
-	// TODO(fabxc): this could potentially pushed further down into the store API to make true streaming possible.
-	sortDedupLabels(resp.seriesSet, q.replicaLabels)
+	if resp.sortRequired {
```

Comment:
I think in the future we should have `q.proxy` have a method that says whether extra sorting is needed or not and then call it before everything to see whether we can stream responses. If not, re-use `seriesServer`. If yes, we could have some kind of adapter returned to Select().

Reply:
One problem I ran into was making this work with the lazy response. I am not how store-gateway sorts series, but if the order is the same as in TSDB, it can send series that cannot be merged on the fly. This can happen if compaction fails for some reason and blocks are not deduplicated.

For example, store-gw can send the following series:
```
a=1, replica=1, b=1
a=1, replica=1, b=2
a=1, replica=2, b=1
```
After reordering labels, we would get
```
a=1, b=1, replica=1
a=1, b=2, replica=1
a=1, b=1, replica=2
```

And this cannot be merged on the fly because `a=1, b=1` will not be deduplicated because the two replicas are not next to each other.

So we will need to properly sort at the source taking replica labels into account, or do some pre-flight checks to know whether a global sort is needed.